zebrad/src/components/inbound.rs

```diff
@@ -52,8 +52,16 @@ pub type SetupData = (Outbound, Arc<Mutex<AddressBook>>);
 /// behind the current tip, while the `Inbound` service is *externally driven*,
 /// responding to block gossip by attempting to download and validate advertised
 /// blocks.
+///
+/// ## Correctness
+///
+/// The `state` service must not return `Poll::Pending`. If it does, a bug in the
+/// `ServiceExt::call_all` implementation might cause the `state` buffer to fill
+/// up, and make Zebra hang.
 pub struct Inbound {
-    // invariant: address_book, outbound, downloads are Some if network_setup is None
+    // invariants:
+    //  * Before setup: address_book and downloads are None, and the *_setup members are Some
+    //  * After setup: address_book and downloads are Some, and the *_setup members are None
```

Comment:
Instead of using two options that alternate between being setup and empty, we should see if we can introduce an enum to represent the two `Option`s together.

```rust
enum VerifierState {
    Initializing(Verifier),
    Initialized(Pin<Box<Downloads...>>),
}
```

Reply:
Yeah, I also had this thought - it's better than trying to reason about this invariant every time we make a change to the Inbound service.

I'd also like to tie all 4 Options together, so we don't have to check the setup invariant at all:
```rust
enum Setup {
    AwaitingNetwork {
        network_setup: oneshot::Receiver<SetupData>,
        verifier: Verifier,
    },
    Initializing,
    Initialized {
        address_book: Arc<Mutex<zn::AddressBook>>,
        downloads: Pin<Box<Downloads...>>),
    },
}
```

I think we need an empty Initializing state so we can:
1. take ownership of the oneshot using swap(setup, Initializing)
2. run some code to fetch the address_book and create the download, consuming the oneshot (and verifier)
3. change the Setup as Initialized using the new address_book and downloads

Otherwise, I think we run into ownership issues with try_recv, because it requires an owned oneshot.

(If we were just wrapping the verifier in an enum, two states would be fine, because we could clone the verifier, then implicitly drop the original when we replaced AwatingNetwork with Initialized.)

Reply:
We probably also want to add address_book_as_ref and downloads_as_mut convenience methods, which return None when we're not Initialized.

That way, the call code stays ergonomic.

Reply:
And finally, we can delete all the top-level and member-specific documentation about invariants, because they're now encoded in the type system.

Reply:
I've finished most of the `Setup` code, but I haven't written any convenience functions yet.

Turns out there were two hidden extra failure states - which I guess isn't really that surprising.

Here are the convenience functions I'm thinking of:
* an `is_* -> bool` for every variant
* `address_book_as_ref -> Option<&Arc<Mutex<zn::AddressBook>>>`
* `downloads_as_mut -> Option<&mut Pin<Box<InboundDownloads>>>`

And maybe these accessors, if they help us avoid destructuring or `mem::swap`:
* `verifier_clone -> Option<&Verifier>`
* `network_setup_as_mut -> Option<&mut oneshot::Receiver<NetworkSetupData>>`
* `failed_setup_as_mut -> Option<&mut oneshot::Receiver<NetworkSetupData>>`

@yaahc do you think this cleanup is important enough to delay merging these improvements?
We could open a ticket and do it later. The code seems functional and readable as it is.

The next sprint is already pretty full for me. I'd like to focus on fixing hangs that are impacting users, rather than cleaning up code that's already decently readable.

Reply:
I don't think the cleanup is particularly important. Like you said the current version is readable and easy to reason about its functionality.